homeassistant/components/mawaqit/__init__.py

```diff
@@ -0,0 +1,501 @@
+            maghrib_hour = maghrib_hour.strftime("%H:%M")
+
+            # isha + 1 minute because this function is launched 1 minute after 'Isha, (useful only if 'Isha is at 11:59 PM)
+            isha_hour = self.prayer_times_info["Isha"] + timedelta(minutes=1)
+            isha_hour = isha_hour.strftime("%H:%M")
+
+            # If 'Isha is before 12 AM (Maghrib hour < 'Isha hour), we need to get the next day's Fajr.
+            # Else, we get the current day's Fajr.
+            if maghrib_hour < isha_hour:
+                index_day = today.day + 1
+            else:
+                index_day = today.day
+
+            try:
+                day_times = month_times[str(index_day)]
+            except KeyError:
+                # If index_day + 1 == 32 (or 31) and the month contains only 31 (or 30) days
+                # We take the first day of the following month (reset 0 if we're in december)
+                if index_month == 11:
+                    index_next_month = 0
+                else:
+                    index_next_month = index_month + 1
+                day_times = calendar[index_next_month]["1"]
+            fajr_hour = day_times[0]
+
+            self.prayer_times_info["Next Salat Name"] = "Fajr"
+            self.prayer_times_info["Next Salat Time"] = dt_util.parse_datetime(
+                f"{today.year}-{today.month}-{index_day} {fajr_hour}:00"
+            )
+
+        countdown_next_prayer = 15
+        if self.prayer_times_info["Next Salat Time"] is not None:
+            self.prayer_times_info["Next Salat Preparation"] = self.prayer_times_info[
+                "Next Salat Time"
+            ] - timedelta(minutes=countdown_next_prayer)
+        else:
+            # TODO check if this is correct # pylint: disable=fixme
+            self.prayer_times_info["Next Salat Preparation"] = None
+
+        _LOGGER.debug("Next salat info updated, updating sensors")
+        async_dispatcher_send(self.hass, DATA_UPDATED)
+
+    async def async_update(self, *_):
+        """Update sensors with new prayer times."""
+        try:
+            # should we use self.hass.async_add_executor_job to wrap get_new_prayer_times?
+            prayer_times = await self.get_new_prayer_times()
+            self.available = True
+        except (BadCredentialsException, ConnError):
+            self.available = False
+            _LOGGER.debug("Error retrieving prayer times")
+            async_call_later(self.hass, 60, self.async_update)
+            return
+
+        _LOGGER.debug("[;] prayer_times : %s", prayer_times)
+        _LOGGER.debug("[;] ha_now() times : %s", ha_now().time())
+        _LOGGER.debug("[;] ha_now() date : %s", ha_now().date())
+
+        for prayer, time in prayer_times.items():
+            tomorrow = (ha_now().date() + timedelta(days=1)).strftime("%Y-%m-%d")
+            today = ha_now().date().strftime("%Y-%m-%d")
+
+            now = ha_now().time().strftime("%H:%M")
```

Comment:
You are doing a lot of time parsing and calculations. These should go into the library and should not be part of HA